cli/src/ui.rs

```diff
@@ -81,13 +89,24 @@ impl BuiltinPager {
         pager
             .set_exit_strategy(minus::ExitStrategy::PagerQuit)
             .expect("Able to set the exit strategy");
-        let pager_handle = pager.clone();
+        let pager_handle = Arc::new(RwLock::new(Some(pager.clone())));
+
+        pager
+            .add_exit_callback(Box::new({
+                let pager_handle = pager_handle.clone();
+                move || {
+                    pager_handle.write().unwrap().take();
```

Comment:
In order to disable late `write()`s, we don't have to drop `pager` object. Instead, maybe we can use `AtomicBool` to be set on pager quit.

BTW, this PR revealed that the current builtin pager is fundamentally broken. Because the backing channel is unbounded, `.write()` will never stuck and burn CPU time.

Reply:
> Because the backing channel is unbounded, .write() will never stuck and burn CPU time.

What do you mean? Are you talking about the problem this PR solves or about a separate problem? Are you talking about something we do or `minus`'s design?

Reply:
Also, AtomicBool sounds fine to me as far as this PR goes, but I'll mention that I was considering replacing the `Option` here with a `Result` later for https://github.com/martinvonz/jj/issues/4182#issuecomment-2259315065. I am not 100% sure that's the best solution to that problem, but it would need preserving more information than just one AtomicBool.

Reply:
It's kinda the same problem under the hood, but user might not notice. Basically, if the pager is minus, `jj log -r.. -p` will scan all ancestors, materialize diffs, and the whole output will be buffered in memory.

Our expectation is that `.write()` into the pager will stuck if the pipe is full, but `minus` doesn't because the underlying channel is unbounded.

Reply:
What would be symptoms of this? Running out of memory? Something more serious?

Would anything short of redesigning minus help? Though, they might accept a PR.

I think the pager Sapling used, streampager, was more carefully designed for huge inputs, but it seemed abandoned last I checked.

Reply:
Wasted CPU time is probably more important. It will surely reach the memory limit, but the output speed wouldn't be that high.

There's a related issue in minus. I guess this wouldn't be an easy issue to fix.
https://github.com/AMythicDev/minus/issues/106

Reply:
@yuja: It's true that we _could_ use `AtomicBool` for signaling here, but I see no advantage to doing so, and a few advantages in keeping the current `Option` container:

1. If we don't intend to use the pager anymore, this makes it inaccessible by design. It matches our intent; I think it will be far less surprising for those unfamiliar with this code to see this handle being discarded than an `AtomicBool` as a peer.
2. Because `minus`' pager handle is implemented as a sender and receiver channel pair for duplix comms. with the paging thread, dropping `pager_handle`'s contents should free the CLI thread's `Receiver`'s resource (which is very relevant to the above discussion about resource usage with lines that haven't been displayed yet). It's unclear to me what comms. might be queued up in return from the paging thread, but this sidesteps any question of whether they're significant.

Reply:
Hmm, I would say `AtomicBool` is simpler and that's the advantage.

As far as I understand the implementation, `minus::dynamic_paging()` doesn't drop their `tx` end, so dropping our `pager` instance wouldn't signal the termination. Ideally, all `rx` ends should be dropped on pager quit, so any further `pager.push_str()` automatically fails. However, `minus` API appears not designed as such use case in mind.

Reply:
Is the point that `AtomicBool` avoids the possibility of deadlocks? While it's more primitive, that makes its docs more complicated (https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html, for example), so I couldn't immediately figure that out.

Like Erich, I liked the idea of making invalid states inexpressible, but I'm not sure about the downsides of RwLock and whether they are worth it.

Reply:
> Is the point that `AtomicBool` avoids the possibility of deadlocks?

No? I just mean boolean flag is easier to follow than heavy lifting on interior mutability.

BTW, judging from the discussion on Discord, it might be better to just switch to streampager.

Reply:
If anyone is reading this, here is what seems like a nice book that discusses Atomic types:

https://marabos.nl/atomics/atomics.html#atomic-load-and-store-operations